### ISSUE TITLE
Start and end position of an error

### ISSUE THREAD

**Issue by katspaugh** [flycheck/flycheck#527]
Hi Sebastian,

I'm very glad you've added the ability to define checkers as generic functions. I've made [a checker for tern-lint](https://gist.github.com/katspaugh/7d646074e547cf59dd6c) based on your flycheck-ocaml example.

The thing is, tern-lint provides error locations as start and end positions, since it uses a JavaScript parser. So I wonder if passing these positions to flycheck could allow it to highlight symbols more accurately then `bounds-of-thing-at-point` (which I assume to be the only way in flycheck to highlight symbols)?

To illustrate, here's how flycheck-tern-lint displays errors in a test file:

![Imgur](http://i.imgur.com/FCES1BQ.png)

For comparison, here's what I get when [using start and end positions directly](https://github.com/katspaugh/tern-lint.el/blob/master/tern-lint.el#L40):

![Imgur](http://i.imgur.com/JsrmcHY.png)

**Comment by jbaum98** [comment]
What's the status of this? I'm working on adding support for Java and I currently have javac giving me file positions for the start and end of the error, and I'd like to put that into flycheck for the most helpful highlighting.

Is this in progress? If not, is there somebody who can give me some direction on what needs to be done? I'd be happy to work on this myself.

**Comment by cpitclaudel** [comment]
Hi @jbaum98!

> What's the status of this? 

I don't think any of us is working on this ATM.  lunaryorn stepped down as maintainer (@fmdkdd and I replaced him).  But! We're still very interested in getting this feature.  It's just that no-one has found the time.

> I'm working on adding support for Java

That would be great!

>is there somebody who can give me some direction on what needs to be done? I'd be happy to work on this myself.

That would be awesome.  It shouldn't be too too hard, but it needs care.  The idea I think is to define a range structure and use that instead of the line/column pairs that we currently have in errors.  Then the code that determines overlay bounds would be moved up to set proper end bounds in cases where the checker returns a single point instead of a range, and the overlay-creation code would be simplified to just use the bounds present in the code.

This would be great to have.  Feel free to open a PR early for code review/discussion as you work on this! :) We'll be very happy to help.

**Comment by jbaum98** [comment]
I've started work on this a little. What should happen to the current highlighting modes? It seems like if we have start and end positions it is reasonable to use as much specificity as the tool give us.

For example:
* If the tool give us just a line (or range of lines), we highlight the whole line(s).
* If the tool gives us a line and a column, we highlight that single character?
* If the tool gives us a start line and a start column, the highlight that range.

So from this it seems like I need to write 5 "constructors" for my `range` struct:
1. `flycheck-range-new`: takes `start-pos` and `end-pos` as keyword arguments.
1. `flycheck-range-for-line`: takes `line` and produces a range covering the whole line.
1. `flycheck-range-for-lines`: takes `start-line` and `end-line` and produces a range covering all the lines.
1. `flycheck-range-for-line-col`: takes `line` and `col` and produces a range covering just the character at (`line`,`col`).
1. `flycheck-range-for-lines-cols`: takes `start-line`, `start-col`, `end-line`, and `end-col` and produces a range from (`start-line`,`start-col`) to (`end-line`,`end-col`).

Does all that sounds reasonable?

**Comment by cpitclaudel** [comment]
My intuition is this: when the tool has extra information that we used to ignore, we should certainly stop ignoring it. In other cases, we should maintain existing behavior.  Concretely:

> If the tool give us just a line (or range of lines), we highlight the whole line(s).

Correct; this is what we currently do.

> If the tool gives us a line and a column, we highlight that single character?
   
I'd prefer maintaining the current behavior, implemented in `flycheck-error-region-for-mode`

> If the tool gives us a start line and a start column, the highlight that range.

Do you mean a start line and column + an end line and column? Then yup, I agree :)

> So from this it seems like I need to write 5 "constructors" for my range struct:

Sounds a bit complicated? My intuition would be to store only "real" data in the struct; there'd be just one constructor that takes start line, end line, start column, end column, and we'd set any missing data to nil.  Then we'd extend flycheck-error-region-for-mode to translate these line+column pairs into buffer offsets (begin and end points), ignoring the value of MODE if all four values are non-nil.  

WDYT?

**Comment by cpitclaudel** [comment]
And of course:

> I've started work on this a little.

🎆 Thanks!

**Comment by jbaum98** [comment]
Thanks for the quick response!

> My intuition would be to store only "real" data in the struct

I agree, I actually think the struct should only store a single start position and end position because those will eventually be the arguments to `make-overlay`. The other "constructors" would be wrappers around the one main constructor and take a buffer so that it can find the correct start and end positions for the struct. They wouldn't be `cl-defstruct` constructors but rather regular elisp functions.

> I'd prefer maintaining the current behavior, implemented in flycheck-error-region-for-mode

Okay, I can see why it would be desirable to maintain current behavior for this case, but it does seem like a little bit much to have the highlighting mode only for the case where the tool provides only line and column and not an end line and column. Are there a lot of tools in this category?

**Comment by jbaum98** [comment]
Maybe to clarify, here is the section I have so far on the range struct:
```elisp
;;; Range data type
(cl-defstruct (flycheck-range (:constructor flycheck-range-new) start-pos end-pos))

(defmacro flycheck-with-buffer (buffer &rest forms)
  "Switch to BUFFER and evaluate FORMS.

If BUFFER is not live, FORMS are not evaluated."
  (declare (indent 1) (debug t))
  `(when (buffer-live-p buffer)
     (with-current-buffer buffer
       (save-restriction
         (save-excursion
           (widen)
           ,@forms)))))

(defun flycheck-line-col-to-pos (line col buffer)
  "Returns the position corresponding to LINE,COL in BUFFER."
  (flycheck-with-buffer buffer
    (goto-line line)
    (move-to-column column)
    (point)))

(defun flycheck-line-start-to-pos (line buffer)
  "Returns the position corresponding to the first non-whitespace character on LINE
 in BUFFER."
  (flycheck-with-buffer buffer
    (goto-char (point-min))
    (forward-line (- line 1))
    ;; We are at the beginning of the line now, so move to the beginning of
    ;; its indentation, similar to `back-to-indentation'
    (let ((end (line-end-position)))
      (skip-syntax-forward " " end)
      (backward-prefix-chars)
      ;; If the current line is empty, include the previous line break
      ;; character(s) to have any region at all.  When called with 0,
      ;; `line-end-position' gives us the end of the previous line
      (if (eolp) (line-end-position 0) (point)))))

(defun flycheck-line-end-to-pos (line buffer)
  "Returns the position corresponding to last character on LINE in BUFFER."
  (flycheck-with-buffer buffer
    (goto-char (point-min))
    (forward-line (- line 1))
    (line-end-position)))

(defun flycheck-range-for-line (line buffer)
  "Returns a range for covering line LINE of BUFFER."
  (flycheck-range-new
   :start-pos (flycheck-line-start-to-pos line buffer)
   :end-pos   (flycheck-line-end-to-pos   line buffer)))

(defun flycheck-range-for-lines (start-line end-line buffer)
  "Returns a range for covering line START-LINE to END-LINE of BUFFER."
  (flycheck-range-new
   :start-pos (flycheck-line-start-to-pos start-line buffer)
   :end-pos   (flycheck-line-end-to-pos   end-line buffer)))

(defun flycheck-range-for-line-col (line col buffer)
  "Returns a range covering only LINE,COL"
  (flycheck-range-new
   :start-pos (flycheck-line-col-to-pos line col       buffer)
   :end-pos   (flycheck-line-col-to-pos line (+ 1 col) buffer)))

(defun flycheck-range-for-lines-cols (start-line start-col end-line end-col buffer)
  "Returns a range starting at START-LINE,START-COL and ending at END-LINE-END-COL"
  (flycheck-range-new
   :start-pos (flycheck-line-col-to-pos start-line start-col buffer)
   :end-pos   (flycheck-line-col-to-pos end-line   end-col   buffer)))
```

**Comment by cpitclaudel** [comment]
> > My intuition would be to store only "real" data in the struct

> I agree, I actually think the struct should only store a single start position and end position because those will eventually be the arguments to make-overlay. The other "constructors" would be wrappers around the one main constructor and take a buffer so that it can find the correct start and end positions for the struct. They wouldn't be cl-defstruct constructors but rather regular elisp functions.

They can be turned into cl-defstruct constructors ­— that's not an issue.  But I don't think we should do the line/column translation on structure creation: it's a costly process, so I prefer delaying it as much as possible (granted, we currently show all errors; but I'd like to be in a good position to adjust overlays dynamically at some point in the future)

> > I'd prefer maintaining the current behavior, implemented in flycheck-error-region-for-mode

> Okay, I can see why it would be desirable to maintain current behavior for this case, but it does seem like a little bit much to have the highlighting mode only for the case where the tool provides only line and column and not an end line and column. Are there a lot of tools in this category?

I'd say most of them :) But I could be wrong ^^

> Maybe to clarify, here is the section I have so far on the range struct:

Thanks for sharing code.  I think there's some overlap with e.g. `flycheck-error-line-region` in there, right?

**Comment by jbaum98** [comment]
Yeah I copied it over because if we're going to find buffer offsets on struct creation we wouldn't need `flycheck-error-line-region`.

> But I don't think we should do the line/column translation on structure creation

Hmm, the reason I wanted to do the translation right away is that then we can store the data on the error in a single unified format, which makes later processing of errors simpler. Another issue is that at I wanted to also support tools that emit raw buffer offsets instead of lines and columns (like the Java tool I was working on before). 

If it is important to delay the translation though, is there a way to do discriminated unions in elisp? Ie. I want to do something like
```hs
data ErrorRegion = Line { line :: Int } 
                 | Lines { startLine :: Int, endLine:: Int } 
                 | LineCol { line:: Int , col :: Int }
                 | LinesCols { startLine :: Int, startCol :: Int,  endLine :: Int, endCol :: Int }
                 | Positions { startPos :: Int, endPos :: Int }
```

**Comment by cpitclaudel** [comment]
> If it is important to delay the translation though, is there a way to do discriminated unions in elisp? 

Yes, more or less; you combine the fact that every type is pointed (everything is a `Maybe`) with the fact that you can do dynamic type-based dispatch.  Here's how I'd model your example below:

```
data ErrorRegion = Line { line :: Int } 
                 | Lines { startLine :: Int, endLine:: Int } 
                 | LineCol { line:: Int , col :: Int }
                 | LinesCols { startLine :: Int, startCol :: Int,  endLine :: Int, endCol :: Int }
```

These all go in a single struct with three `nil`-able fields.  It over-approximates, but that's OK.

```
                 | Positions { startPos :: Int, endPos :: Int }
```

That one goes in a different struct.

But there's a more important point here: even if you have a checker that returns only offsets, you'll still need to compute line and columns, for e.g. the error list.

Given this, I'd tend to make a single struct with 4 fields for line and column begin and end, and two fields for buffer offsets.  Then I'd make two accessors, on for the line/column endpoints, and one for the buffer endpoints.  The line/column one would use offsets to compute lines and columns if needed, and vice-versa, and both would store the computed results for faster future access.

WDYT?

**Comment by jbaum98** [comment]
We need line and column numbers for error messages?

Okay sounds good. Just to make sure I understand how this will play out in each case
* Line(s): creates a struct with only start line and end line (maybe set to the same thing). Accessor lazily calculates buffer offsets but columns stay nil.
* Line and column: Creates struct with line and column but end line and column nil. Buffer offsets are calculated lazily according to current highlighting mode.
* Line and column for start and end: sets all 4 start and end lines and columns. Lazily calculates buffer offsets in accessor.
* Start and end position: creates struct with start and end positions, lazily calculates line and col with accessor.

**Comment by cpitclaudel** [comment]
> We need line and column numbers for error messages?

We show them in the error list, yup.

> Just to make sure I understand how this will play out in each case (…)

Yes, I think that sounds pretty good to me, so I'd say go ahead if that looks good to you too!  I'm sure @fmdkdd will also have comments when we have an implementation :)

**Comment by fmdkdd** [comment]
> I'm sure @fmdkdd will also have comments when we have an implementation :)

I remember I hacked `flycheck-error-region-for-mode` to make it work with `start/end` columns, but then I never submitted a patch because it was obvious that larger changes would be required, since the code relied too much on a single line/column coordinate.  

But I fully agree that we should use the ranges provided by the checkers, if they are present.  Using a `range` struct seems like a good way to do it.

Thanks @jbaum98 for tackling this one!

**Comment by david-christiansen** [comment]
Yes, thank you very much! I'll be keeping an eye on this, as I have a few tools that provide accurate source spans where Flycheck would save a lot of work over hand-rolling something.

**Comment by cpitclaudel** [comment]
This is implemented now :) @david-christiansen , it would be great if you could have a look at https://github.com/flycheck/flycheck/pull/1684

**Comment by david-christiansen** [comment]
I'm on 60% parental leave, so I don't really have time to look. But it
sounds good! Thanks!

**Comment by cpitclaudel** [comment]
Congratulations!
